contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1727,6 +1730,204 @@ object_name(PG_FUNCTION_ARGS)
+
+            if (!user)
+            {
+                pfree(db_name);
+                pfree(schema_name);
+                pfree(object_name);
+                PG_RETURN_NULL();
+            }
+            else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+            {
+                physical_schema_name = pstrdup(get_guest_schema_name(db_name));
+            }
+            else
+            {
+                pfree(schema_name);
+                schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+                physical_schema_name = get_physical_schema_name(db_name, schema_name);
+            }
+        }
+        else
+        {
+            pfree(db_name);
+            pfree(schema_name);
+            pfree(object_name);
+            PG_RETURN_INT32(result);
+        }
+    }
+    else
+    {
+        physical_schema_name = get_physical_schema_name(db_name, schema_name);
+    }
+
+    /* get schema oid from physical schema name, it will return InvalidOid if user don't have lookup access */
+    schema_oid = get_namespace_oid(physical_schema_name, true);
+    pfree(db_name);
+    pfree(physical_schema_name);
+
+    if (!OidIsValid(schema_oid) || pg_namespace_aclcheck(schema_oid, user_id, ACL_USAGE) != ACLCHECK_OK)
+    {
+        pfree(schema_name);
+        pfree(object_name);
+        PG_RETURN_NULL();
+    }
+    // If schema is 'sys' or 'pg_catalog' then search in typecode list, else in pg_type.
+    if(!strcmp(schema_name, "sys") || !strcmp(schema_name, "pg_catalog"))
+    {
+        result = (*common_utility_plugin_ptr->get_tsql_datatype_oid) (object_name);
+    }
+    else
+    {
+        result = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid,
+                                    CStringGetDatum(object_name), ObjectIdGetDatum(schema_oid));
+    }
+    pfree(schema_name);
+    pfree(object_name);
+    if (OidIsValid(result) && pg_type_aclcheck(result, user_id, ACL_USAGE) == ACLCHECK_OK)
+        PG_RETURN_INT32(result);
+    else
+        PG_RETURN_NULL();
+}
+
+/*
+ * type_name
+ *      returns the type name with type id as input
+ * Returns NULL
+ *      if there is no such type
+ *      if user don't have right permission
+ */
+Datum
+type_name(PG_FUNCTION_ARGS)
+{
+    Datum       type_id = PG_GETARG_DATUM(0);
+    Datum       tsql_typename;
+    HeapTuple   tuple;
+    Oid         user_id = GetUserId();
+    char        *result = NULL;
+
+    LOCAL_FCINFO(fcinfo1, 1);
+
+    if (type_id < 0)
+        PG_RETURN_NULL();
+    
+    InitFunctionCallInfoData(*fcinfo1, NULL, 1, InvalidOid, NULL, NULL);
+    fcinfo1->args[0].value = type_id;
+    fcinfo1->args[0].isnull = false;
+    // Search in typecode list, it will return type name if system datatype, else will return null.
+    tsql_typename = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo1);
+    if (tsql_typename)
+    {
+        PG_RETURN_VARCHAR_P(tsql_typename);
```

Comment:
Let us return PG_RETURN_TEXT_P instead of PG_RETURN_VARCHAR_P as we have marked the return type of the corresponding sql function as NVARCHAR. Similar signature can be found in the implementation of user_name, suser_name etc. Similarly update the return type to PG_RETURN_TEXT_P for all the c functions whose corresponding sql function's return type is NVARCHAR